### ISSUE TITLE
Difficulty obtaining meaningful syllables & applying a trained model to new Drosophila datasets

### ISSUE THREAD

**Issue by anarac97** [dattalab/keypoint-moseq#267]
Hello!

I'm currently exploring keypoint-MoSeq for behavioral analysis in Drosophila melanogaster and would really appreciate some guidance.

I'm using keypoint-MoSeq to identify stereotyped behaviors in freely walking flies across different experimental conditions and timepoints. The flies I'm recording walk in a conical arena and are recorded from below. 

My dataset consists of 4 experimental conditions, 10 videos per condition (~120,000 frames per video, 900x900, 100fps). For pose estimation, I trained two independent SLEAP models: 1) for my control condition, I trained a mode with 17 keypoints that is working fairly well (average distance (ground truth vs prediction): 1.29 px, mAP: 0.72, mAR: 0.79) and 2) for my experimental conditions, I trained a model with 13 keypoints (the same keypoints as above, except that 4 are absent), that still makes some mislabelling (average distance (ground truth vs prediction): 1.79 px, mAP: 0.49, mAR: 0.58)

I first focused on my control condition using the 17-keypoint SLEAP tracking data to determine whether keypoint-MoSeq could identify distinct behavioral syllables (I'm following the publicly available Colab tutorial: https://keypoint-moseq.readthedocs.io/en/latest/). I used essentially the default parameters, including kappa = 1e3, that you team used to segment fly behavior in the keypoint MoSeq paper. 

My main issue is behavioral segmentation. Although the model identifies approximately 20 syllables, the resulting segmentation is much less specific than I expected. By manually inspecting my videos, I can clearly distinguish several different walking behaviors and other subtle movement patterns. However, after fitting the model, I can only recognize only two or three broad behavioral categories (walking, turning and idle), most syllables appear to contain multiple distinct behaviors and, when inspecting the grid movies, I frequently find that different behavioral motifs are grouped into the same syllable, making the segmentation difficult to interpret.

Because of this, I'm unsure whether this reflects limitations of my tracking quality, my dataset or whether I should be tuning different model parameters besides kappa. So, I'd appreciate any suggestions regarding recommended parameters to tune (I've played around with kappa and I haven't seen any major improvements), whether this behavior is expected for Drosophila walking or whether there are common pitfalls that could explain these results.

I'm also struggling on running the model in Google Colab. Even when using a GPU, I repeatedly encounter memory errors during fitting. Is there a recommended strategy for datasets of this size? 

Finally, I have a question regarding model transfer. Suppose I successfully train a model using my control dataset (17 keypoints). Would it be appropriate to apply this trained model to my experimental datasets, which contain only 13 keypoints or would you instead recommend retraining a new model using only the common 13 keypoints across all datasets? 

I'd be very grateful for any recommendations on how to make better use of keypoint-MoSeq for this application. I'm also happy to share example videos, SLEAP tracking files, or additional details of my pipeline if that would be helpful.

Thank you very much!

**Comment by hatifnatt57** [comment]
Hello!

I am a fellow researcher that studies birds and am in no way connected to the development of Keypoint-Moseq. I personally had much better outcomes after running the location-aware version of the model, passing the location_aware=True argument as recommended here:
https://keypoint-moseq.readthedocs.io/en/latest/advanced.html#location-aware-modeling
Without it the syllables I was getting weren't very meaningful, not only for my data, but also for the example dataset that is provided in the tutorial notebook.

In my experience the location-aware model needs a larger kappa to obtain the same median syllable duration (1e6 for the regular model resulted in the same median duration as 1e10 in the location-aware one).

Hope that helps in some way.

**Comment by JRJacoby** [comment]
Hi @anarac97, sorry for the slow reply, and thanks for the detailed writeup.

The exact kappa that works is going to be different for each dataset. The metric to watch is the median syllable duration you get from each kappa. You want a median syllable duration that comes out around the length of the behaviors that you're interested in. We found that with flies, we got only a few very general behaviors at longer timescales, with more specific behaviors coming out at shorter timescales. What median syllable durations are you getting with each kappa you try? Are they close to the duration of your behaviors of interest?

I'm not used to fly datasets but latent_dim = 10 sounds high to me. Are you setting this according to the variance explained metric from the PCA initialization step? If so, can you share the visualization of the PCs and the scree plot? Do the PCs look like identifiable fly movements to you? 

One of the problems people run into using kpms with non-rodent species is that some of the speed priors aren't sensible for other species. Older versions of kpms used to hard-code this and newer versions have a notebook cell in the pipeline for estimating hyperparams. What version are you running? 

Anecdotally, a lot of users do find they like the location_aware version of the model better. It's worth a shot. 

You won't be able to fit 4.5M frames into Google Colab. I'd recommend training on a stratified subsample of your sessions and then using apply_model to apply that one model in batches to the rest of the dataset. 

You'll have to train on the 13 shared keypoints.
